data_structures/src/chain/priority.rs

```diff
@@ -0,0 +1,1337 @@
+    }
+}
+
+/// Updates an `Option` with the value from another candidate `Option` always that the old `Option`
+/// was `None` or contained a value that was greater than the candidate.
+///
+/// This differs from `std::cmp::min` in the case that the existing `Option` is `None`, because
+/// `impl Ord for Option` considers that `None` is always smaller than `Some(_)`:
+///
+/// ```
+/// use witnet_data_structures::chain::priority::Priority;
+///
+/// let mut option = None;
+/// option = std::cmp::min(option, Some(Priority::from(2337)));
+/// option = std::cmp::min(option, Some(Priority::from(1337)));
+/// option = std::cmp::min(option, Some(Priority::from(3337)));
+///
+/// assert_eq!(option, None);
+/// ```
+///
+/// ```
+/// use witnet_data_structures::chain::priority::{option_update_if_less_than, Priority};
+///
+/// let mut option = None;
+/// option_update_if_less_than(&mut option, Some(Priority::from(2337)));
+/// option_update_if_less_than(&mut option, Some(Priority::from(1337)));
+/// option_update_if_less_than(&mut option, Some(Priority::from(3337)));
+///
+/// assert_eq!(option, Some(Priority::from(1337)))
+/// ```
+#[inline]
+pub fn option_update_if_less_than(option: &mut Option<Priority>, candidate: Option<Priority>) {
+    match (&candidate, &option) {
+        (Some(new), Some(old)) if new < old => {
+            *option = candidate;
+        }
+        (Some(_), None) => {
+            *option = candidate;
+        }
+        _ => {}
+    }
+}
+
+/// Priority estimation strategies. To be used with `PriorityEngine::estimate_priority`.
+pub mod strategies {
+    use pdatastructs::topk::lossycounter::LossyCounter;
+
+    use super::*;
+
+    /// A priority estimation strategy that receives a list of targetted time-to-blocks expressed in
+    /// minutes and derives the priority from there by using a [lossy count algorithm].
+    ///
+    /// [lossy count algorithm]: https://en.wikipedia.org/wiki/Lossy_Count_Algorithm
+    #[allow(clippy::cast_precision_loss)]
+    #[allow(clippy::cast_possible_truncation)]
+    #[allow(clippy::cast_sign_loss)]
+    pub fn target_minutes<'a, I>(
+        priorities: I,
+        target_minutes: [u16; 5],
+        seconds_per_epoch: u16,
+    ) -> PrioritiesEstimate
+    where
+        I: IntoIterator<Item = &'a Priorities>,
+        I::IntoIter: Clone + ExactSizeIterator,
+    {
+        // Make the priorities argument an iterator (if it was not already) and measure its length.
+        let priorities = priorities.into_iter();
+        let priorities_count = priorities.len() as f64;
+
+        // Set the number of buckets used to ease the frequency counting
+        let buckets_count = 50.0;
+
+        // Create lossy counters for measuring frequency of priorities separately for DRTs and VTTs
+        // As we are working on small data sets, we can use a near-zero epsilon value, which
+        // essentially neutralizes the "lossiness" of the lossy counters.
+        let epsilon = 0.0000000001;
```

Comment:
Does this mean that we could use a normal counter instead?

Reply:
Hmmm actually yes, we should be able to replace `LossyCounter` with a simpler counter, as we're not making the most of the losiness and performing bucketing outside of it anyway.

Reply:
However, `LossyCounter` implements querying by frequency, which is needed for this use case and I would prefer not to implement myself.

Reply:
E.g. if we use something like `counter::Counter`, we would need to turn the counter into a `HashMap`, then fold that into a vector of frequencies (alike to a histogram), and then iterate over that to find the lowest buckets whose frequencies are higher than each of the queried frequencies.

Reply:
I gave it a try and this commit passes all the tests:

https://github.com/witnet/witnet-rust/commit/04aa61befd5418100f6ae057b402262dd6c88b0d

Reply:
Thanks, I cherry-picked that into my branch!

Reply:
Nice! However I forgot to remove the dependency from `data_structures/Cargo.toml`